src/ftest.jl

```diff
@@ -229,3 +229,204 @@ function show(io::IO, ftr::FTestResult{N}) where N
     end
     print(io, '─'^totwidth)
 end
+
+
+##############################################
+# Tests of Between-Subjects Effects
+# Baset on F-statistics
+# L: The s×p full row rank matrix. The rows are estimable functions. s≥1 where p number of coefs
+
+struct GroupEffectsTable
+    name::Vector{String}
+    df::Vector{Float64}
+    f::Vector{Float64}
+    p::Vector{Float64}
+end
+
+"""
+θ + A * B * A'
+
+Change θ (only upper triangle). B is symmetric.
+"""
+function mulαβαtinc!(θ::AbstractMatrix, A::AbstractMatrix, B::AbstractMatrix)
+    axb  = axes(B, 1)
+    sa   = size(A, 1)
+    for m ∈ 1:sa
+        for n ∈ m:sa
+            for j ∈ axb
+                @inbounds for i ∈ axb
+                    θ[m, n] +=  A[m, i] * B[i, j] * A[n, j]
+                end
+            end
+        end
+    end
+    θ
+end
+"""
+a' * B * a
+
+"""
+function mulαtβα(a::AbstractVector, B::AbstractMatrix{T}) where T
+    if length(a) != size(B, 2)::Int  || size(B, 2)::Int  != size(B, 1)::Int  error("Dimention error") end
+    axbm  = axes(B, 1)
+    axbn  = axes(B, 2)
+    c = zero(T)
+    for i ∈ axbm
+        ct = zero(T)
+        for j ∈ axbn
+            @inbounds  ct += B[i, j] * a[j]
+        end
+        @inbounds c += ct * a[i]
+    end
+    c
+end
+
+# See SPSS (GLM/UNIANOVA) and SAS (PROC GLM) documentation 
+# https://www.ibm.com/docs/en/spss-statistics/29.0.0?topic=effects-tests-between-subjects
+# L is a s×p matrix corresponding to plan-matrix of Factor
+# p - number of columns - coefs number
+# s - number of levels for this factor in the model
+# For Example
+# If you have model matrix with Intercept and two factors A and B with 3 and 4 levels
+# with Dummy coding you will have:
+# 
+# I A2 A3 B2 B3 B4
+# 1 1  0  1  0  0
+# 1 1  0  1  0  0
+# 1 1  0  1  0  0
+# 1 1  0  1  0  0
+# 1 0  1  1  0  0
+# 1 0  1  0  1  0
+# 1 0  1  0  1  0
+# 1 0  1  0  1  0
+# 1 0  1  0  1  0
+# 1 0  0  0  0  1
+# 1 0  0  0  0  1
+# 1 0  0  0  0  1
+# 1 0  0  0  0  0
+# 1 0  0  0  0  0
+#
+# Then yoy wil have L matrix fo intercept:
+#
+# 1 0  0  0  0  0 
+#
+# For A:
+#
+# 0 1  0  0  0  0
+# 0 0  1  0  0  0
+#
+# For B:
+#
+# 0 0  0  1  0  0
+# 0 0  0  0  1  0
+# 0 0  0  0  0  1
+#
+"""
+    lcontrast(obj, i::Int)
+
+L-contrast matrix for `i` fixed effect.
+"""
+function lcontrast(obj, i::Int)
+    n = length(obj.mf.f.rhs.terms)
```

Comment:
Much of this function relies _very_ heavily on the internal structures of the types from StatsModels. If any internal structures change in StatsModels (which is typically not considered a breaking change in the Julia world) this function will break. StatsModels provides a variety of functions with methods to access these quantities in a stable way; this function should be refactored to use those.

Reply:
Sorry, I can't find public API from StatsModel documentation to deep work with terms.  Can you kindly point me for that ways...